### DISCUSSION TITLE
How to calculate depth and complexity of a query in instrumentation?

### DISCUSSION THREAD

**Discussion by justinsumida** [graphql-java/graphql-java#3632] [Q&A]
We have an Instrumentation class that logs information about the Graph query/mutation happening. Some things that we log include the complexity and depth of the query.
With recent upgrades to 21.3, we can't seem to "hijack" the call to MaxQueryComplexity/DepthInstrumentation to get that. When I tried to extract that code that was called in beginValidation(), the instrumentation fails to move on.

Just was curious if others have successfully calculated depth/complexity before.

**Comment by justinsumida** [comment]
I should back this up with:
It **does** work on normal queries/mutations.
It **does not** work when the query/mutation had some errors that the GraphQL-Java picked up before getting to our logging instrumentation (for example, missing a NonNull input, failing @OneOf, etc...)

Is there a reason it doesn't work when there are existing ValidationErrors before getting into beingValidation()?

**Comment by bbakerman** [comment]
in 18.1 we changed the MaxQueryComplexity / MaxQueryDepthInstrumentation to run at `beginExecuteOperation` for stability reasons.

https://github.com/graphql-java/graphql-java/pull/2773

Its a more sensible time - the query is valid and ready to be executed so now you can calculate complexity and depth for example.

I am not sure what version you are migrating from but they now do their work starting at `beginExecuteOperation`

**Reply by justinsumida** [reply]
We upgraded from 17.3, so that would explain why. And we were seeing the exact behavior you explained in that PR. We were getting a "should have been validated before" error when we upgraded and had null/non null mismatches on inputs.
Do you suggest taking the depth and complexity calculations from those instrumentations to run our own calculations to then log them?

**Comment by bbakerman** [comment]
> Do you suggest taking the depth and complexity calculations from those instrumentations to run our own calculations to then log them?

yes we have teams doing this at the company I work at (Atlassian)

Rememebr you can "tweak" the complexity calculation yoursrelf via your own callback and hence you could look at directives say and calc cost from them (imagine a `@expensive(add : 100)` directive that added more cost to the calculation say)
